File: src/test/java/com/fwdekker/randomness/integer/IntegerSettingsTest.java
package com.fwdekker.randomness.integer;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;


/**
 * Unit tests for {@link IntegerSettings}.
 */
final class IntegerSettingsTest {
    private IntegerSettings integerSettings;


    @BeforeEach
    void beforeEach() {
        integerSettings = new IntegerSettings();
    }


    @Test
    void testGetLoadState() {
        integerSettings.setMinValue(742);
        integerSettings.setMaxValue(908);
        integerSettings.setBase(12);

        final IntegerSettings newIntegerSettings = new IntegerSettings();
        newIntegerSettings.loadState(integerSettings.getState());

        assertThat(newIntegerSettings.getMinValue()).isEqualTo(742);
        assertThat(newIntegerSettings.getMaxValue()).isEqualTo(908);
        assertThat(newIntegerSettings.getBase()).isEqualTo(12);
    }

    @Test
    void testGetSetMinValue() {
        integerSettings.setMinValue(366);

        assertThat(integerSettings.getMinValue()).isEqualTo(366);
    }

    @Test
    void testGetSetMaxValue() {
        integerSettings.setMaxValue(332);

        assertThat(integerSettings.getMaxValue()).isEqualTo(332);
    }

    @Test
    void testGetSetBase() {
        integerSettings.setBase(7);

        assertThat(integerSettings.getBase()).isEqualTo(7);
    }

    @Test
    void testGetSetGroupingSeparator() {
        integerSettings.setGroupingSeparator('6');

        assertThat(integerSettings.getGroupingSeparator()).isEqualTo('6');
    }

    @Test
    void testGetSetGroupingSeparatorStringEmpty() {
        integerSettings.setGroupingSeparator("");

        assertThat(integerSettings.getGroupingSeparator()).isEqualTo('\0');
    }

    @Test
    void testGetSetGroupingSeparatorString() {
        integerSettings.setGroupingSeparator("tlRg}");

        assertThat(integerSettings.getGroupingSeparator()).isEqualTo('t');
    }
}

File: src/test/kotlin/com/fwdekker/randomness/integer/IntegerSettingsTest.kt
package com.fwdekker.randomness.integer

import org.assertj.core.api.Assertions.assertThat
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe


/**
 * Unit tests for [IntegerSettings].
 */
object IntegerSettingsTest : Spek({
    lateinit var integerSettings: IntegerSettings


    beforeEachTest {
        integerSettings = IntegerSettings()
    }


    describe("copying") {
        it("creates an independent copy") {
            val copy = integerSettings.deepCopy()

            integerSettings.currentScheme.minValue = 159
            copy.currentScheme.minValue = 48

            assertThat(integerSettings.currentScheme.minValue).isEqualTo(159)
        }

        it("copies state from another instance") {
            integerSettings.currentScheme.minValue = 742
            integerSettings.currentScheme.maxValue = 908
            integerSettings.currentScheme.base = 12

            val newIntegerSettings = IntegerSettings()
            newIntegerSettings.loadState(integerSettings.state)

            assertThat(newIntegerSettings.currentScheme.minValue).isEqualTo(742)
            assertThat(newIntegerSettings.currentScheme.maxValue).isEqualTo(908)
            assertThat(newIntegerSettings.currentScheme.base).isEqualTo(12)
        }
    }
})


/**
 * Unit tests for [IntegerScheme].
 */
object IntegerSchemeTest : Spek({
    lateinit var integerScheme: IntegerScheme


    beforeEachTest {
        integerScheme = IntegerScheme()
    }


    describe("input handling") {
        describe("grouping separator") {
            it("uses the default separator if null is set") {
                integerScheme.safeSetGroupingSeparator(null)

                assertThat(integerScheme.groupingSeparator).isEqualTo(IntegerScheme.DEFAULT_GROUPING_SEPARATOR)
            }

            it("uses the default separator if an empty string is set") {
                integerScheme.safeSetGroupingSeparator("")

                assertThat(integerScheme.groupingSeparator).isEqualTo(IntegerScheme.DEFAULT_GROUPING_SEPARATOR)
            }

            it("uses only the first character if a multi-character string is given") {
                integerScheme.safeSetGroupingSeparator("recited")

                assertThat(integerScheme.groupingSeparator).isEqualTo("r")
            }
        }
    }

    describe("copying") {
        describe("copyFrom") {
            it("makes the two schemes equal") {
                val schemeA = IntegerScheme()
                val schemeB = IntegerScheme(myName = "Name")
                assertThat(schemeA).isNotEqualTo(schemeB)

                schemeA.copyFrom(schemeB)

                assertThat(schemeA).isEqualTo(schemeB)
            }
        }

        describe("copyAs") {
            it("makes two schemes equal except for the name") {
                val schemeA = IntegerScheme()
                val schemeB = schemeA.copyAs("NewName")
                assertThat(schemeA).isNotEqualTo(schemeB)

                schemeB.myName = schemeA.myName

                assertThat(schemeA).isEqualTo(schemeB)
            }
        }
    }
})
